From 412b8c8bf2ad020aa4014a4666b3b4f0d98aea37 Mon Sep 17 00:00:00 2001
From: De Huo <De.Huo@windriver.com>
Date: Fri, 19 Jul 2019 15:23:50 +0800
Subject: [PATCH] Fix rule accounts_password_pam_unix_remember remediate error

Fix remediate error problem for rule accounts_password_pam_unix_remember.

Upstream-Status: Pending

Signed-off-by: De Huo <De.Huo@windriver.com>
---
 .../bash/wrlinux1019.sh                            | 11 +++++
 .../oval/wrlinux1019.xml                           | 49 ++++++++++++++++++++++
 2 files changed, 60 insertions(+)
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_unix_remember/bash/wrlinux1019.sh
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_unix_remember/oval/wrlinux1019.xml

diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_unix_remember/bash/wrlinux1019.sh b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_unix_remember/bash/wrlinux1019.sh
new file mode 100644
index 0000000..42a0bd1
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_unix_remember/bash/wrlinux1019.sh
@@ -0,0 +1,11 @@
+# platform = multi_platform_wrlinux
+. /usr/share/scap-security-guide/remediation_functions
+populate var_password_pam_unix_remember
+
+pamFile="/etc/pam.d/common-password"
+
+if grep -q "remember=" $pamFile; then
+	sed -i --follow-symlinks "s/\(^password.*sufficient.*pam_unix.so.*\)\(\(remember *= *\)[^ $]*\)/\1remember=$var_password_pam_unix_remember/" $pamFile
+else
+	sed -i --follow-symlinks "/^password[[:space:]]\+.*\+pam_unix.so/ s/$/ remember=$var_password_pam_unix_remember/" $pamFile
+fi
diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_unix_remember/oval/wrlinux1019.xml b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_unix_remember/oval/wrlinux1019.xml
new file mode 100644
index 0000000..c2e6c33
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_password_pam_unix_remember/oval/wrlinux1019.xml
@@ -0,0 +1,49 @@
+<def-group>
+  <definition class="compliance" id="accounts_password_pam_unix_remember" version="2">
+    <metadata>
+      <title>Limit Password Reuse</title>
+      <affected family="unix">
+        <platform>multi_platform_wrlinux</platform>
+      </affected>
+      <description>The passwords to remember should be set correctly.</description>
+    </metadata>
+    <criteria comment="remember parameter of pam_unix.so or pam_pwhistory.so is set correctly" operator="OR">
+      <criterion comment="remember parameter of pam_unix.so is set correctly" test_ref="test_accounts_password_pam_unix_remember" />
+      <criterion comment="remember parameter of pam_pwhistory.so is set correctly" test_ref="test_accounts_password_pam_pwhistory_remember" />
+    </criteria>
+  </definition>
+
+  <!-- Check the pam_unix.so remember case -->
+  <ind:textfilecontent54_test id="test_accounts_password_pam_unix_remember" check="all" check_existence="all_exist"
+  comment="Test if remember attribute of pam_unix.so is set correctly in /etc/pam.d/common-password" version="1">
+    <ind:object object_ref="object_accounts_password_pam_unix_remember" />
+    <ind:state state_ref="state_accounts_password_pam_unix_remember" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="object_accounts_password_pam_unix_remember" version="1">
+    <ind:filepath>/etc/pam.d/common-password</ind:filepath>
+    <ind:pattern operation="pattern match">^\s*password\s+(?:(?:sufficient)|(?:required)|(?:\[success=.*default=(?:(?:bad)|(?:ignore))\]))\s+pam_unix\.so.*remember=(\d+).*$</ind:pattern>
+    <ind:instance datatype="int">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+  <!-- Check the pam_pwhistory.so remember case -->
+  <ind:textfilecontent54_test id="test_accounts_password_pam_pwhistory_remember" check="all" check_existence="all_exist"
+  comment="Test if remember attribute of pam_pwhistory.so is set correctly in /etc/pam.d/common-password" version="1">
+    <ind:object object_ref="object_accounts_password_pam_pwhistory_remember" />
+    <ind:state state_ref="state_accounts_password_pam_unix_remember" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="object_accounts_password_pam_pwhistory_remember" version="1">
+    <ind:filepath>/etc/pam.d/common-password</ind:filepath>
+    <ind:pattern operation="pattern match">^\s*password\s+(?:(?:requisite)|(?:required))\s+pam_pwhistory\.so.*remember=([0-9]*).*$</ind:pattern>
+    <ind:instance datatype="int">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+  <!-- Common state - shared by both of 'remember' tests above -->
+  <ind:textfilecontent54_state id="state_accounts_password_pam_unix_remember" version="1">
+    <ind:subexpression datatype="int" operation="greater than or equal" var_ref="var_password_pam_unix_remember" />
+  </ind:textfilecontent54_state>
+
+  <external_variable comment="number of passwords that should be remembered" datatype="int" id="var_password_pam_unix_remember" version="1" />
+
+</def-group>
-- 
1.9.1

